### ISSUE TITLE
Allow gather and indexing ops to return sparse gradients

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#16329]
## 🚀 Feature
Currently, autograd engine can handle a mixture of sparse and dense gradients, but to the best of my knowledge only embedding has an option to return sparse gradients. Yet sparse gradients also naturally arise in gather/indexing ops, so it would be good to have an option for those ops to return sparse gradients

## Motivation

1) Consider label-smoothed criterion in nlp workloads. One of the formulations (the particular form is not important) is
```loss = a * logprobs.mean(-1)+(1-a)*logprobs.gather(dim=-1, index=target.unsqueeze(1)).squeeze(1)```
By far the biggest bottleneck for backward is dense gradient accumulation, where 2 reads and 1 write of the dense tensor(s) of `batch_size x vocab_size` tensors is happening, and before that dense gradient for `gather` was materialized, mostly consisting of zeros, whereas all that is really needed is adding sparse gradients from `gather` to the logprobs.grad
2) We had a few issues recently about `nan` gradients for values that don't even contribute to loss
https://github.com/pytorch/pytorch/issues/15506
https://github.com/pytorch/pytorch/issues/15131
with the root cause summarized by @t-vi https://github.com/pytorch/pytorch/issues/15506#issuecomment-450239626
"The problem is that index select will give zero gradient for the non-selected indices.
But you then multiply by NaN when taking the log derivative.
There is no way to distinguish "0 gradient from not used" to 0 from "the gradient happens to be 0". Then you arrive at "NaN * 0". The NaN in the gradient is inserted by the derivative formula of "log" not of the indexing."
Returning sparse gradients will allow one to distinguish between "0 gradient from not used" (there will be no gradient in the sparse case) and "the gradient happens to be 0". 
There's a question of how far sparse gradients should be propagated back, but as a first step, just returning sparse gradients and immediately converting them to dense/accumulating them to dense is enough to solve problem 1) and at least some varieties of problem 2). Also, it's not clear how to indicate sparse gradients for advanced indexing operations on the API level. 

## Pitch

Have `gather` and `index_select` and similar operations support `sparse = True/False` argument with the default value of `False`

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @carlosgmartin, @t-vi, @vishwakftw, @samedii, @mruberry

**Comment by apaszke** [comment]
FWIW wider support for sparse gradients wouldn't resolve your second point completely. It might eliminate the issue in some cases, but I'm pretty sure there will still be many failure modes.

**Comment by yf225** [comment]
cc @gchanan
